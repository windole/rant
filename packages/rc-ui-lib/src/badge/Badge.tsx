import React, { CSSProperties, useContext } from 'react';
import classnames from 'classnames';
import { BadgeProps } from './PropsType';
import { isDef, addUnit } from '../utils';
import { isNumeric } from '../utils/validate/number';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';

const Badge: React.FC<BadgeProps> = (props) => {
  const { content, max, dot, showZero, tag } = props;

  const { prefixCls, createNamespace } = useContext(ConfigProviderContext);
  const [bem] = createNamespace('badge', prefixCls);

  const hasContent = () => {
    // if (props.content) {
    //   return true;
    // }
    return isDef(content) && content !== '' && (showZero || +content !== 0);
  };

  const renderContent = () => {
    if (!dot && hasContent()) {
      if (isDef(max) && isNumeric(content?.toString()) && +content > max) {
        return `${max}+`;
      }

      return content;
    }
    return null;
  };

  const renderBadge = () => {
    if (hasContent() || props.dot) {
      let style: CSSProperties = {
        background: props.color,
      };

      if (props.offset) {
        const [x, y] = props.offset;

        if (props.children) {
          style.top = addUnit(y);
          if (typeof x === 'number') {
            style.right = addUnit(-x);
          } else {
            style.right = x.startsWith('-') ? x.replace('-', '') : `-${x}`;
          }
        } else {
          style.marginTop = addUnit(y);
          style.marginLeft = addUnit(x);
        }
      }

      if (!props.children) {
        style = { ...props.style, ...style };
      }
      return (
        <div
          className={classnames(
            {
              [props.className]: props.className && !props.children,
            },
            bem({ dot: props.dot, fixed: !!props.children }),
          )}
          style={style}
        >
          {renderContent()}
        </div>
      );
    }
    return null;
  };

  if (props.children) {
    return React.createElement(
      tag,
      {
        className: classnames(bem('wrapper'), props.className),
        onClick: props.onClick,
        onTouchStart: props.onTouchStart,
        style: props.style,
      },
      props.children,
      renderBadge(),
    );
  }

  return renderBadge();
};

Badge.defaultProps = {
  tag: 'div',
  showZero: true,
};

export default Badge;
