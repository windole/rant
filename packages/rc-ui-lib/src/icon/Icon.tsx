import React, { useContext, useMemo } from 'react';
import classnames from 'classnames';
import { addUnit } from '../utils';
import { IconProps } from './PropsType';
import Badge from '../badge';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';

function isImage(name?: string): boolean {
  return name ? name.indexOf('/') !== -1 : false;
}

const Icon: React.FC<IconProps> = (props) => {
  const { iconPrefix } = useContext(ConfigProviderContext);

  const { dot, tag, name, className, onClick, onTouchStart } = props;

  const imageIcon = isImage(name);

  const classPrefix = useMemo(
    () => props.classPrefix || iconPrefix,
    [props.classPrefix, iconPrefix],
  );

  return (
    <Badge
      dot={dot}
      tag={tag}
      className={classnames(className, classPrefix, imageIcon ? '' : `${classPrefix}-${name}`)}
      style={{
        color: props.color,
        fontSize: addUnit(props.size),
        ...props.style,
      }}
      onClick={onClick}
      onTouchStart={onTouchStart}
      {...props.badge}
    >
      {props?.children}
      {imageIcon && <img className={classnames('rc-icon__image')} src={name} alt={name} />}
    </Badge>
  );
};

Icon.defaultProps = {
  tag: 'i',
};

export default Icon;
