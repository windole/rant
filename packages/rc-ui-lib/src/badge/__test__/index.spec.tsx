import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Badge from '../index';
import Icon from '../../icon';

it('should render nothing when content is empty string', () => {
  const wrapper = mount(<Badge content="" />);

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('should render nothing when content is undefined', () => {
  const wrapper = mount(<Badge content="undefined" />);

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('should render nothing when content is zero', () => {
  const wrapper = mount(<Badge content="0" />);

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('should render content slot correctly', () => {
  const wrapper = mount(
    <Badge content={<Icon name="success" className="badge-icon" />}>
      <div className="child" />
    </Badge>,
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('should change dot position when using offset prop', async () => {
  const wrapper = mount(
    <Badge dot offset={[2, 4]}>
      <div className="child" />
    </Badge>,
  );

  const badge = wrapper.find('.rc-badge');
  expect(badge.props().style.top).toEqual('4px');
  expect(badge.props().style.right).toEqual('-2px');

  await wrapper.setProps({
    offset: [-2, -4],
  });
  expect(wrapper.find('.rc-badge').props().style.top).toEqual('-4px');
  expect(wrapper.find('.rc-badge').props().style.right).toEqual('2px');
});

it('should change dot position when using offset prop with custom unit', async () => {
  const wrapper = mount(
    <Badge dot offset={['2rem', '4em']}>
      <div className="child" />
    </Badge>,
  );

  const badge = wrapper.find('.rc-badge');
  expect(badge.getDOMNode().style.top).toEqual('4em');
  expect(badge.getDOMNode().style.right).toEqual('-2rem');

  await wrapper.setProps({
    offset: ['-2rem', '-4em'],
  });

  expect(wrapper.find('.rc-badge').props().style.top).toEqual('-4em');
  expect(wrapper.find('.rc-badge').props().style.right).toEqual('2rem');
});

it('should change dot position when using offset prop without children', () => {
  const wrapper = mount(<Badge dot offset={[2, 4]} />);

  const badge = wrapper.find('.rc-badge');
  expect(badge.props().style.marginTop).toEqual('4px');
  expect(badge.props().style.marginLeft).toEqual('2px');
});

it('should not render zero when show-zero is false', async () => {
  const wrapper = mount(
    <Badge content="0">
      <div className="child" />
    </Badge>,
  );

  expect(wrapper.find('.rc-badge').exists()).toBeTruthy();
  await wrapper.setProps({ showZero: false });
  expect(wrapper.find('.rc-badge').exists()).toEqual(false);
});

it('can set custom max count', () => {
  const wrapper = mount(
    <Badge content={100} max={99}>
      news
    </Badge>,
  );
  expect(wrapper.find('.rc-badge').text()).toBe('99+');
});
