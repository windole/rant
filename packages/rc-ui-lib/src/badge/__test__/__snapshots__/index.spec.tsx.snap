// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render content slot correctly 1`] = `
<Badge
  content={
    <Icon
      className="badge-icon"
      name="success"
      tag="i"
    />
  }
  showZero={true}
  tag="div"
>
  <div
    className="rc-badge__wrapper"
  >
    <div
      className="child"
    />
    <div
      className="rc-badge rc-badge--fixed"
      style={
        Object {
          "background": undefined,
        }
      }
    >
      <Icon
        className="badge-icon"
        name="success"
        tag="i"
      >
        <Badge
          className="badge-icon van-icon van-icon-success"
          showZero={true}
          style={
            Object {
              "color": undefined,
              "fontSize": undefined,
            }
          }
          tag="i"
        >
          <i
            className="rc-badge__wrapper badge-icon van-icon van-icon-success"
            style={
              Object {
                "color": undefined,
                "fontSize": undefined,
              }
            }
          />
        </Badge>
      </Icon>
    </div>
  </div>
</Badge>
`;

exports[`should render nothing when content is empty string 1`] = `
<Badge
  content=""
  showZero={true}
  tag="div"
/>
`;

exports[`should render nothing when content is undefined 1`] = `
<Badge
  content="undefined"
  showZero={true}
  tag="div"
>
  <div
    className="rc-badge"
    style={
      Object {
        "background": undefined,
      }
    }
  >
    undefined
  </div>
</Badge>
`;

exports[`should render nothing when content is zero 1`] = `
<Badge
  content="0"
  showZero={true}
  tag="div"
>
  <div
    className="rc-badge"
    style={
      Object {
        "background": undefined,
      }
    }
  >
    0
  </div>
</Badge>
`;
